x/superfluid/keeper/stake.go

```diff
@@ -287,6 +287,48 @@ func (k Keeper) undelegateCommon(ctx sdk.Context, sender string, lockID uint64)
 	return intermediaryAcc, nil
 }
 
+// partialSuperfluidUndelegate acts similarly to undelegateCommon, but undelegates a partial amount of the lock's delegation rather than the full amount. The amount
+// that is undelegated is placed in a new lock. This function returns the intermediary account associated with the original lock ID as well as the new lock that was created.
+// An error is returned if the amount to undelegate is greater than the locked amount.
+func (k Keeper) partialSuperfluidUndelegate(ctx sdk.Context, sender string, lockID uint64, amountToUndelegate sdk.Coin) (intermediaryAcc types.SuperfluidIntermediaryAccount, newlock *lockuptypes.PeriodLock, err error) {
+	lock, err := k.lk.GetLockByID(ctx, lockID)
+	if err != nil {
+		return types.SuperfluidIntermediaryAccount{}, &lockuptypes.PeriodLock{}, err
+	}
+	err = k.validateLockForSF(ctx, lock, sender)
+	if err != nil {
+		return types.SuperfluidIntermediaryAccount{}, &lockuptypes.PeriodLock{}, err
+	}
+
+	if amountToUndelegate.Amount.GTE(lock.Coins[0].Amount) {
+		return types.SuperfluidIntermediaryAccount{}, &lockuptypes.PeriodLock{}, fmt.Errorf("partial undelegate amount must be less than the locked amount")
+	}
+
+	// get the intermediate account associated with lock id, and delete the connection.
+	intermediaryAcc, found := k.GetIntermediaryAccountFromLockId(ctx, lockID)
+	if !found {
+		return types.SuperfluidIntermediaryAccount{}, &lockuptypes.PeriodLock{}, types.ErrNotSuperfluidUsedLockup
+	}
+
+	// undelegate the desired lock amount, and burn the minted osmo.
+	amount, err := k.GetSuperfluidOSMOTokens(ctx, intermediaryAcc.Denom, amountToUndelegate.Amount)
+	if err != nil {
+		return types.SuperfluidIntermediaryAccount{}, &lockuptypes.PeriodLock{}, err
+	}
+	err = k.forceUndelegateAndBurnOsmoTokens(ctx, amount, intermediaryAcc)
+	if err != nil {
+		return types.SuperfluidIntermediaryAccount{}, &lockuptypes.PeriodLock{}, err
+	}
+
+	// Move the funds from the old lock to a new lock with the remaining amount.
+	newLock, err := k.lk.SplitLock(ctx, *lock, sdk.NewCoins(amountToUndelegate), true)
+	if err != nil {
+		return types.SuperfluidIntermediaryAccount{}, &lockuptypes.PeriodLock{}, err
```

Comment:
coming to think about it, do we not need logic to handle the lock after calling `forceUndelegateAndBurnOsmoTokens`? Things that come to my mind are
- lock id and intermediary account connection
- handle  synthethic lock for the undelegating lock
- create synthlock for the unbonding amounts

Reply:
Depending on the asnwer for this, could we have according checks added for each of ^?

Reply:
I am not sure which side of the action you are referring to.

If you are referring to the remaining amount in the lock that still exists:
- the lockID and intermediary account connection remains the same
- the synthetic lock for that lock also remains the same
- the underlying lock of the synthetic lock is reduced, so that handles the synth lock portion

If you are referring to the new amount the got put into a lock that got migrated:
- the lockID and intermediary account connection and synthetic lock is made in the migration function itself (see https://github.com/osmosis-labs/osmosis/blob/efb468ce574e910214a6cdad93d237b4ad786a77/x/superfluid/keeper/migrate.go#L150-L163)

Reply:
This is a good catch, but I think the logic of this function needs to stay as it is for the following reason:

We don't to create all those connections just to remove those connections so we can migrate it. That means this function is to be used by CL migration only. That being said, two things need to be done:

- This function should be renamed and a warning given that connections are not made after the lock is split
- I can abstract this into a helper function to a real partialSuperfluidUndelegate function that DOES create the connections, but just not call that function and instead call the helper directly for this

Again, nice call on this!

Reply:
So ended up making a shared function, I think you will agree with the implementation https://github.com/osmosis-labs/osmosis/pull/5162/commits/9a911853bb8b881df39a7577be321ff8d05a52d5

(still using the same logic as before, but if someone wants to make a partialSuperfluidUndelegate message in the future, they wont accidentally use the wrong method)